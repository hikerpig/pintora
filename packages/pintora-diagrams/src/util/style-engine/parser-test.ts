import { type ActionHandler, type MakeAction } from '../base-db'
import { genParserWithRules } from '../parser-util'
import { StyleRule, type StyleSelector } from './shared'
import grammar from './grammar/style-test'
import { StylePayloads, IStyleDb } from './parser'

const parseFn = genParserWithRules(grammar, {
  dedupeAmbigousResults: true,
})

export type ParserStyleRule = {
  selector: StyleSelector
  attrs: Array<{
    key: string
    value: string
  }>
}

export type Action = MakeAction<StylePayloads>

export type StyleData = {
  styleRules: StyleRule[]
}

export const STYLE_ACTION_HANDLERS: {
  [K in keyof StylePayloads]: ActionHandler<StylePayloads, IStyleDb, K>
} = {
  style(action) {
    // 处理添加样式的逻辑
    const rules: StyleRule[] = []
    action.rules.forEach(rule => {
      rules.push({
        selector: rule.selector,
        attrs: rule.attrs.reduce((previous, current) => {
          previous[current.key] = current.value
          return previous
        }, {}),
      })
    })
    this.styleRules.push(...rules)
  },
}

export class StyleDb {
  styleRules: StyleRule[] = []

  private ACTION_HANDLERS = STYLE_ACTION_HANDLERS

  apply(action: Action | Action[]) {
    if (!action) return
    if (Array.isArray(action)) {
      action.forEach(a => this.apply(a))
      return
    }
    if (action.type in this.ACTION_HANDLERS) {
      this.ACTION_HANDLERS[action.type].call(this, action as any)
    }
  }

  getData(): StyleData {
    return {
      styleRules: this.styleRules,
    }
  }
}

export class StyleParse {
  static parse(text: string) {
    const actions = parseFn(text)
    const styleDb = new StyleDb()
    styleDb.apply(actions)
    return styleDb.getData()
  }
}
