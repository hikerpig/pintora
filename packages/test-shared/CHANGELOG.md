# Change Log

## 0.2.0-alpha.3

### Patch Changes

- bc16cae: [gantt] add support for axisInterval
- 46a1a9f: [gantt] add support for axisInterval

## 0.2.0-alpha.2

### Patch Changes

- 235cc0e: add @config and @param to gantt diagram; update tmLanguage;

## 0.2.0-alpha.1

### Minor Changes

- bd214c7: Add gantt diagram

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2-beta.0](https://github.com/hikerpig/pintora/compare/v0.1.1...v0.1.2-beta.0) (2022-02-13)

### Bug Fixes

- should support `\r` as end-of-line char and whitespace char ([f0696f4](https://github.com/hikerpig/pintora/commit/f0696f48dcb2e82b7368b8a2908dc938b6c3aea2))

# [0.1.0](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.27...v0.1.0) (2022-02-06)

**Note:** Version bump only for package @pintora/test-shared

# [0.1.0-alpha.24](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.23...v0.1.0-alpha.24) (2022-02-01)

### Features

- **pintora-diagrams:** [componentDiagram] we can have relationship from element to group ([9cc74cd](https://github.com/hikerpig/pintora/commit/9cc74cdc53013b6b63a13658ff719d414d04706f))

# [0.1.0-alpha.19](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.18...v0.1.0-alpha.19) (2022-01-22)

### Features

- **pintora-diagrams:** add mindmap support ([3062216](https://github.com/hikerpig/pintora/commit/306221660b34baad83d67435470d6aabc27b289f))

# [0.1.0-alpha.18](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.17...v0.1.0-alpha.18) (2022-01-16)

### Features

- **pintora-cli:** read cli args `--theme` ([a502c35](https://github.com/hikerpig/pintora/commit/a502c35cf1dbbad65c5847e2752e17711b8dccef))

# [0.1.0-alpha.11](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.10...v0.1.0-alpha.11) (2021-12-04)

### Features

- **pintora-diagrams:** Add basic activityDiagram [#3](https://github.com/hikerpig/pintora/issues/3) ([718a80e](https://github.com/hikerpig/pintora/commit/718a80e55d209d618fb855a463a151b880bf6fc3))

# [0.1.0-alpha.8](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.7...v0.1.0-alpha.8) (2021-08-29)

### Features

- [sequenceDiagram] add custom group color ([f93fe8a](https://github.com/hikerpig/pintora/commit/f93fe8a6fdb81af55eeb6f303f426c6150eeaf86))

# [0.1.0-alpha.4](https://github.com/hikerpig/pintora/compare/v0.1.0-alpha.3...v0.1.0-alpha.4) (2021-08-15)

### Bug Fixes

- **pintora-diagrams:** [componentDiagram] avoid interface node overlapping ([213cea0](https://github.com/hikerpig/pintora/commit/213cea090f40717390417e143366026d841f9808))

### Features

- **pintora-diagrams:** [erDiagram] add support for key coloumn like 'PK' ([64dad3d](https://github.com/hikerpig/pintora/commit/64dad3ddd8cefb912d7c83e14cf8a2dc0e64c5f8))

# [0.1.0-alpha.0](https://github.com/hikerpig/pintora/compare/v0.0.1...v0.1.0-alpha.0) (2021-08-01)

### Features

- Add componentDiagram ([c967f1c](https://github.com/hikerpig/pintora/commit/c967f1c9a969a813adac9a9967a06a024f04d73f))
