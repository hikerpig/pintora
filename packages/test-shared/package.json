{
  "name": "@pintora/test-shared",
  "version": "0.2.0-alpha.3",
  "description": "Shared testing utils for pintora",
  "keywords": [
    "pintora"
  ],
  "author": "hikerpig <hikerpigwinnie@gmail.com>",
  "homepage": "https://github.com/hikerpig/pintora#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "sideEffects": false,
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/hikerpig/pintora.git"
  },
  "scripts": {
    "compile": "tsc",
    "watch": "tsc -w"
  },
  "bugs": {
    "url": "https://github.com/hikerpig/pintora/issues"
  },
  "gitHead": "92814f05b3df654b588fe35e3af275a60ea61296"
}
