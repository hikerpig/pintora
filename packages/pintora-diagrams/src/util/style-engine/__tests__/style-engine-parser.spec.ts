import { StyleParse } from '../parser-test'
import { stripStartEmptyLines } from '@pintora/test-shared'

describe('style engine parser', () => {
  it('can parse @style block', () => {
    const example = stripStartEmptyLines(`
@style {
  #Alice {
    fontFamily: serif;
  }
  .message__text {
    textColor: #ff0000;
  }
}
    `)
    const result = StyleParse.parse(example)
    expect(result).toMatchSnapshot()
  })

  it('can parse bindClass', () => {
    const example = stripStartEmptyLines(`
    @bindClass e1,e2 test-class
    `)
    const result = StyleParse.parse(example)
    expect(result.bindRules[0]).toMatchObject({
      nodes: ['e1', 'e2'],
      className: 'test-class',
    })
  })
})
