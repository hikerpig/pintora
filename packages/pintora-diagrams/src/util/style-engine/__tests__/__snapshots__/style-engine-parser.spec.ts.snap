// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style engine parser can parse @style block 1`] = `
{
  "bindRules": [],
  "configParams": [],
  "overrideConfig": {},
  "styleRules": [
    {
      "attrs": {
        "fontFamily": "serif",
      },
      "selector": {
        "target": "Alice",
        "type": "id",
      },
    },
    {
      "attrs": {
        "textColor": "#ff0000",
      },
      "selector": {
        "target": "message__text",
        "type": "class",
      },
    },
  ],
  "title": "",
}
`;
